### PR TITLE
Topic recognition - S4745193 Stable Diffusion on ADNI

### PR DESCRIPTION
**Pull request by SixLeopard**
Hey Teaching Team,

This is a pull request the ingrates a solution for generating new MRI scans for AD or NC based on the ADNI dataset using stable diffusion. The goal was to create reasonably good looking brain images with style/conditioning where during generation you can enter whether you want to generate AD or NC images. I Believe this goal has been achieved as the model supports the label conditioning and can supply during generation and it generated reasonably good looking brain images for the type you entered.

Sample of generated Images:

Files Added
```
47451933_Stable_Diffusion
├── Results (train results + generation outputs + diagrams)
│   ├── BrainSeq1.png
│   ├── Brains1.png
│   ├── COMP3710SD_diagram.png
│   ├── COMP3710SD_diagram_infrence.png
│   ├── Diffusion_loss.png
│   ├── Diffusion_loss_2.png
│   ├── Diffusion_loss_21.png
│   ├── Diffusion_loss_22.png
│   ├── Diffusion_loss_23.png
│   ├── Diffusion_loss_24.png
│   ├── UMapPlot.png
│   ├── VAE_loss.png
│   ├── VAE_loss_2.png
│   ├── VAE_loss_21.png
│   └── VAE_loss_22.png
│   └── AD_Genrated_Brains.png
│   └── NC_Genrated_Brains.png
│   └── test_brains.png
├── .gitignore
├── README.md
├── dataset.py (for loading in the ADNI dataset, train,test,validation splits, loading transformations)
├── modules.py (contains all the networks)
├── predict.py (reads the trained models from file and uses them to generate completely new images from noise)
├── project_setup.py (use to setup folder structure and check dataset is available)
├── requirements.txt (what python libraries re required)
├── train.py (trains the model, both the vae and U-net, can change settings to output intermediate results and save the models to file)
└── utils.py (contains extra functions used by most the files)
```

Documentation
The ReadMe.md contains all the information about the model and the results including example generated outputs, loss curves from training and Umap visualization of the latent space.

Kind Regards,
Jamie Westerhout
s4745193

### CODE REVIEW

**Comment by gayanku** [comment]
*****************************************************************
**This is an initial inspection, no action is required at this point**
Stable Diffusion ADNI brain data ---- Hard Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 4.5 | Conditional. OK images, perhaps more time steps?	|																					|
| Implementation functions as intended	| 3 | 2.5 | |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 5 | Hard Difficulty |
| Section IV : Max mark 20 from 20		|   | **19**   | |

**TSNE / UMAP:** 
Good, clear separation. 

**Suggestions:**
Might be good to have mote time steps to improve image clarity.
Good Setup/Reproducibility/Requirements would be good to have.

**Comment by SixLeopard** [comment]
i have updated the "Reproducing Results" section to be better explain the Setup/Reproducibility/Requirements. Tried increasing the number time steps but didn't create a significant improvement in image clarity.

**Comment by SixLeopard** [comment]
reverted back to version originally submitted based on ed thread #504

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Images need more clarity.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Stable Diffusion (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Minor typos and issues. Usage needs more content.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by YashMitttal** [comment]
Some additional feedback:

- Incorporating metrics like Frechet Inception Distance (FID) or structural similarity index (SSIM) to validate the generated images' quality quantitatively would have been helpful.
- It would help to include more information on specific hyperparameters used during training. Provide a rationale for the choice of hyperparameters
- The VAE training loss graph shows signs of overfitting. To address this, you might consider adding regularization methods like dropout in the VAE or employing an early stopping criterion.
- While the training and validation splits are well-handled, expanding on the testing methodology would add rigor to the evaluation.

Overall, good job!

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Some overfitting in VAE.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. Stable Diffusion (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. FID/SSIM metrics needed.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
